---
title: "oculina16s_analysis"
output:
  word_document: default
  html_document: default
  pdf_document: default
date: '2022-06-01'
editor_options: 
  chunk_output_type: console
---
### Based on DADA2 Pipeline 1.16 Walkthrough & Nicola Kriefall's Moorea holobiont analysis

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup

## Packages
```{r packages}
library(ggplot2)
library(cowplot)
library(phyloseq)
library(car)
library(ggpubr)
library(vegan)
library(dada2)
library(stats)
library(MuMIn)
library(sjPlot)
library(webshot)
library(FSA)
library(flextable)
```



## Read in data

```{r read in data}
#setwd("~/oculina/data/")
samdf <- read.csv("~/oculina/data/oculina16s_sampledata_symdens.csv",header=TRUE) #R is making me put the full path all of a sudden, not sure why
row.names(samdf) <- samdf$id
load("~/oculina/data/taxa2.rev.Rdata")
```

## Phyloseq objects

```{r making phyloseq objects, eval=FALSE}
seqtab.rare <- read.csv("~/oculina/data/oculina16s_rev_seqtab.rare.2k",row.names=1)
ps.rare <- phyloseq(otu_table(seqtab.rare, taxa_are_rows=FALSE), 
                    sample_data(samdf), 
                    tax_table(taxa2))
ps.rare  
#save(ps.rare,file="ps.rare.rev.Rdata")
#5138 taxa - raw & rarefied


```

##TRIMMED VERSIONS:
```{r}
seqtab.rare.trim <- read.csv("oculina16s_rev_seqtab.trim.rare.2k.csv",row.names=1)
ps.rare.trim <- phyloseq(otu_table(seqtab.rare.trim, taxa_are_rows=FALSE), 
                    sample_data(samdf), 
                    tax_table(taxa2))
ps.rare.trim
#save(ps.rare.trim,file="ps.rare.trim.rev.Rdata")
#823 taxa &  64 samples rarefied

#checking what happens if we read in the unrarefied version:
seqtab.trim <- read.csv("oculina16s_rev_seqtab.trim.csv",row.names=1)
ps.trim <- phyloseq(otu_table(seqtab.trim, taxa_are_rows=FALSE), 
                    sample_data(samdf), 
                    tax_table(taxa2))
ps.trim
#save(ps.trim,file="ps.trim.rev.Rdata")
#823 taxa & 67 samples not rarefied
```


```{r read in ps objects}
setwd("~/oculina/data")
#load("ps.clean.Rdata")
load("~/oculina/data/ps.rare.rev.Rdata")
#load("ps.rare.trim.Rdata")
#load("ps.trim.Rdata")
```

# Diversity

[Notes from phyloseq author](https://rdrr.io/bioc/phyloseq/man/estimate_richness.html)
Visualize alpha-diversity - Should be done on raw, untrimmed dataset

```{r generate div metrics}
df <- data.frame(estimate_richness(ps.rare, split=TRUE, measures=c("Shannon","InvSimpson","Observed","Chao1")))
#df <- data.frame(estimate_richness(ps.clean, split=TRUE, measures=c("Shannon","InvSimpson","Observed")))
#df <- data.frame(estimate_richness(ps.trim, split=TRUE, measures=c("Shannon","InvSimpson","Observed")))
#df <- data.frame(estimate_richness(ps.rare.trim, split=TRUE, measures=c("Shannon","InvSimpson","Observed")))
df$id <- rownames(df)
df.div <- merge(df,samdf,by="id") #add sample data
#shannon diversity divided by species richness
df.div$even <- df.div$Shannon/(log(df.div$Observed))
df.div$season <- as.factor(df.div$season)
df.div$site <- as.factor(df.div$site)
df.div$type <- as.factor(df.div$type)
df.div$season <- relevel(df.div$season, ref="Summer")
#df.div$site <- relevel(df.div$site, ref="NR")
df.div$site <- factor(df.div$site, levels=c("NR","RI","CL"))

df.div.coral <- subset(df.div, type=="coral")
df.div.NR <- subset(df.div,site=="NR")
df.div.RI <- subset(df.div,site=="RI")
df.div.CL <- subset(df.div,site=="CL")
df.div.summer <- subset(df.div,season=="Summer")
df.div.fall <- subset(df.div,season=="Fall")
df.div.ss <- subset(df.div, type=="sediment" | type =="seawater")
df.div.coral.NR <- subset(df.div.coral,site=="NR")
df.div.coral.RI <- subset(df.div.coral,site=="RI")
df.div.coral.CL <- subset(df.div.coral,site=="CL")
df.div.coral.NR.RI <- subset(df.div.coral,site %in% c("NR","RI"))
df.div.coral$density <- as.numeric(df.div.coral$density)
#skewness(df.div.coral$density) #very right skewed (1.93)
df.div.coral$density_sqrt <- sqrt(df.div.coral$density)


```

```{r Summary statistics all types}
library(gtsummary)

# all types
df1 <- df.div %>% select(Shannon, even, Observed, Chao1, type)
tbl.all.site <- tbl_summary(df1, by = type) %>%
  add_p(test = Shannon ~ "aov") %>% # test for a difference between groups
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()

aov.sha <- (aov(Shannon ~ type, data=df1))
TukeyHSD(aov.sha)
dunnTest(even~type, data=df1, method="bonferroni") #coral-sed, sw-sed sig
dunnTest(Observed~type, data=df1, method="bonferroni") #all sig
dunnTest(Chao1~type, data=df1, method="bonferroni") #coral-sed sig

tbl.all.site %>%
  as_flex_table() %>%
  save_as_image(path ="~/oculina/diversity/figures/tbl.all.site.png") # use extensions .html .tex .ltx .rtf

```

```{r summary statistics coral}


df1 <- df.div.coral %>% select(Shannon, even, Observed, Chao1, site)
tbl.coral.site <- tbl_summary(df1, by = site) %>%
  add_p(test = Shannon ~ "aov") %>% # test for a difference between groups
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()
aov.sha <- (aov(Shannon ~ site, data=df1))
TukeyHSD(aov.sha)

#post hoc test for kruskall wallis is the dunn test
#library(FSA)
dunnTest(even~site, data=df1, method="bonferroni") #CL-RI sig dif, others not

df2 <- df.div.coral %>% select(Shannon, even, Observed, Chao1, season)
tbl.coral.seas <- tbl_summary(df2, by = season) %>%
  add_p() %>% # test for a difference between groups
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()

df3 <- df.div.coral %>% select(Shannon, even, Observed, Chao1, symbstate)
tbl.coral.ss <- tbl_summary(df3, by = symbstate) %>%
  add_p() %>% # test for a difference between groups
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()

tbl.coral.merge <-
  tbl_merge(
    tbls = list(tbl.coral.site, tbl.coral.seas, tbl.coral.ss),
    tab_spanner = c("**Site**", "**Season**", "**Symbiotic State**")
  )

tbl.coral.merge %>%
  as_flex_table() %>%
  save_as_image(path ="~/oculina/diversity/figures/tbl.alphadivstats.coral.merge.png") # use extensions .html .tex .ltx .rtf

```

### Cross validation?
```{r}
library(modelr)
library(purrr)
cv <- crossv_kfold(df.div.coral, k=3)
mod1 <- map(cv$train, ~lm(Shannon ~ site + season + density_sqrt, data=.))
mod2 <- map(cv$train, ~lm(Shannon ~ site + season + symbstate, data=.))

get_pred <- function(model, test_data){
  data <- as.data.frame(test_data)
  pred <- add_predictions(data, model)
  return(pred)
}

pred1 <- map2_df(mod1, cv$test, get_pred, .id= "Run")
pred2 <- map2_df(mod2, cv$test, get_pred, .id= "Run")

MSE1 <- pred1 %>% group_by(Run) %>%
  summarise(MSE=mean(Shannon-pred)^2)

MSE2 <- pred2 %>% group_by(Run) %>%
  summarise(MSE=mean(Shannon-pred)^2)

mean(MSE1$MSE)
mean(MSE2$MSE)


```



### Specificity analysis for symbiont density
```{r}
#library("remotes")
#install_github("darcyj/specificity")
library(specificity)
#need to subset just coral samples from ps.rare
ps.coral <- subset_samples(ps.rare, type=="coral")
#removing sample E2 for now b/c it has NA for density
ps.coral <- subset_samples(ps.coral, id != "E2")
otutable <- as(otu_table(ps.coral), "matrix")
if(taxa_are_rows(ps.coral)){otutable <- t(otutable)}
otutable.df = as.data.frame(otutable)
prop_abund(otutable.df)

#apply occupancy threshold to remove low-occupancy species
otutable.over10 <- occ_threshold(otutable.df, threshold=10)
#how many species left?
ncol(otutable)
ncol(otutable.over10) #50 - a LOT were removed! (started with 5237); common for microbiome data, apparently

#double check that we didn't modify the rows (samples) of otutable
all(rownames(otutable.over10) == rownames(metadata))
#TRUE

# set number of CPU cores to use - change for your system (4 cores for typical hardware)
spec_ncores <- 4

#make empty list for specificity values
specs_list <- list()

###specificity for symb density:
metadata <- sample_data(ps.coral) %>%
  data.frame() %>%
  select("fullname", "density") %>%
  na.omit()
metadata$density <- as.numeric(metadata$density)
specs_list$"SymbDensity" <- phy_or_env_spec(otutable.over10, env=metadata$density,n_sim=1000,n_cores=spec_ncores)

plot_specs_violin(specs_list, label_cex=0.6, cols="red") #all nonsig! lol

#specs_df <- aggregate_specs_list(specs_list, byFeature=FALSE, fd=seqtab.rare, fd_id=1)
#ggplot(specs_list, aes(x=Variable, y=Spec, fill=Variable)) +
#  geom_violin()+
#  geom_jitter(width=0.2,size=0.2)

```


### GLMS - Alpha Diversity Indices

```{r compare types (coral, sed, sw)}

str(df.div)

glm.0 <- glm(Shannon ~ 1, data=df.div, family=gaussian())
glm.1 <- glm(Shannon ~ type, data=df.div, family=gaussian)
glm.2 <- glm(Shannon ~ type + season + site, data=df.div, family=gaussian)
glm.3 <- glm(Shannon ~ type + season * site, data=df.div, family=gaussian)
glm.4 <- glm(Shannon ~ type * season + site, data=df.div, family=gaussian)
glm.5 <- glm(Shannon ~ type * season * site, data=df.div, family=gaussian)

AICc(glm.0, glm.1, glm.2, glm.3, glm.4, glm.5) #AICc 161.69
#glm.1 is best model
summary(glm.1)
#           Estimate Std. Error t value Pr(>|t|)    
#(Intercept)    3.5004     0.1197  29.242  < 2e-16 ***
#typeseawater   0.2379     0.2462   0.966    0.338    
#typesediment   1.5806     0.2539   6.225 4.17e-08 ***

#calc diff in chi square statistics for null model and the model with independent variances (deviances) and dif in df
1-pchisq(61.967-38.515, 66-64)
#8.080959e-06
#reject null hypothesis that there is no different b/w null and model with variables

tab_model(glm.1) #makes a nice looking table
tab_model(glm.1, file="./stats/sha.types.glm.gauss.html")

#cross validation
#library(boot)
cv.err.10 <- cv.glm(df.div, glm.1, K=10)$delta
muhat <- fitted(glm.1)
diag <- glm.diag(glm.1)
(cv.err <- mean((glm.1$y - muhat)^2/(1 - diag$h)^2)) #0.628

### Evenness - all types

glm.0 <- glm(even ~ 1, data=df.div, family=gaussian)
glm.1 <- glm(even ~ type, data=df.div, family=gaussian)
glm.2 <- glm(even ~ type + season + site, data=df.div, family=gaussian)
glm.3 <- glm(even ~ type + season * site, data=df.div, family=gaussian)
glm.4 <- glm(even ~ type * season + site, data=df.div, family=gaussian)
glm.5 <- glm(even ~ type * season * site, data=df.div, family=gaussian)

AICc(glm.0, glm.1, glm.2, glm.3, glm.4, glm.5)
# glm.4 best, AIC -144.84
summary(glm.4)
#                         Estimate Std. Error t value Pr(>|t|)    
#(Intercept)                0.81560    0.03153  25.870  < 2e-16 ***
#typeseawater              -0.14189    0.03951  -3.592 0.000670 ***
#typesediment               0.03497    0.03951   0.885 0.379709    
#seasonSummer              -0.10766    0.02611  -4.123 0.000119 ***
#siteNR                     0.05411    0.02636   2.052 0.044581 *  
#siteRI                     0.03957    0.02622   1.509 0.136605    
#typeseawater:seasonSummer  0.18882    0.04947   3.817 0.000327 ***
#typesediment:seasonSummer  0.12165    0.05122   2.375 0.020804 *  

tab_model(glm.4)
### Richness - all types
library(MASS)

glm.0 <- glm.nb(Observed ~ 1, data=df.div)
glm.1 <- glm.nb(Observed ~ type, data=df.div)
glm.2 <- glm.nb(Observed ~ type + season + site, data=df.div)
glm.3 <- glm.nb(Observed ~ type + season * site, data=df.div)
glm.4 <- glm.nb(Observed ~ type * season + site, data=df.div)
glm.5 <- glm.nb(Observed ~ type * season * site, data=df.div)

AICc(glm.0, glm.1, glm.2, glm.3, glm.4, glm.5)
#glm.1 best (dif from raw analysis, where glm.2 was best), AICc 778.5
summary(glm.1)

#             Estimate Std. Error t value Pr(>|t|)    
#(Intercept)    104.81      17.41   6.019 9.40e-08 ***
#typeseawater    54.58      35.82   1.524    0.133    
#typesediment   256.61      36.94   6.947 2.31e-09 ***

tab_model(glm.1)


### Chao1 - all types

glm.0 <- glm(Chao1 ~ 1, data=df.div, family=gaussian)
glm.1 <- glm(Chao1 ~ type, data=df.div, family=gaussian)
glm.2 <- glm(Chao1 ~ type + season + site, data=df.div, family=gaussian)
glm.3 <- glm(Chao1 ~ type + season * site, data=df.div, family=gaussian)
glm.4 <- glm(Chao1 ~ type * season + site, data=df.div, family=gaussian)
glm.5 <- glm(Chao1 ~ type * season * site, data=df.div, family=gaussian)

AICc(glm.0, glm.1, glm.2, glm.3, glm.4, glm.5)
# glm.1 best, AIC 886.0
summary(glm.1)
tab_model(glm.1)
```

```{r shannon revised stats (glm)}

# Notes on alpha diversity indices:
    # Shannon - measures richness, so how many different bacteria in each sample
    # Chao1 - richness can be sensitive to read depth; Chao1 has some corrections to overcome this
    # Simpson - measures probability of resampling the same species on two consecutive draws with replacement


df.div.coral$season <- as.factor(df.div.coral$season)
df.div.coral$season <- relevel(df.div.coral$season, ref="Summer")
df.div.coral$site <- as.factor(df.div.coral$site)
#df.div.coral$site <- relevel(df.div.coral$site, ref="NR")
df.div.coral$site <- factor(df.div.coral$site, levels=c("NR","RI","CL"))

#df.div.coral$density <- as.numeric(df.div.coral$density)
#df.div.coral$density_sqrt <- sqrt(df.div.coral$density)
df.div.coral$density_sqrt <- as.numeric(df.div.coral$density_sqrt)

library(stats)
#install.packages("MuMIn")
library(MuMIn)
#library(lmer)
set.seed(1)

glm.0 <- glm(Shannon ~ 1, data=df.div.coral, family=gaussian())
glm.1a <- glm(Shannon ~ site, data=df.div.coral, family=gaussian())
glm.1b <- glm(Shannon ~ season, data=df.div.coral, family=gaussian())
glm.1c <- glm(Shannon ~ density_sqrt, data=df.div.coral, family=gaussian())

glm.2a <- glm(Shannon ~ season+site, data=df.div.coral, family=gaussian())
glm.2b <- glm(Shannon ~ season+density_sqrt, data=df.div.coral, family=gaussian())
glm.2c <- glm(Shannon ~ site+density_sqrt, data=df.div.coral, family=gaussian())

glm.3a <- glm(Shannon ~ season+site+density_sqrt, data=df.div.coral, family=gaussian())
glm.3b <- glm(Shannon ~ season+site*density_sqrt, data=df.div.coral, family=gaussian())
glm.3c <- glm(Shannon ~ season*site+density_sqrt, data=df.div.coral, family=gaussian())
glm.3d <- glm(Shannon ~ site+season*density_sqrt, data=df.div.coral, family=gaussian())
glm.3e <- glm(Shannon ~ season*site*density_sqrt, data=df.div.coral, family=gaussian())

AICc(glm.0,glm.1a,glm.1b,glm.1c, glm.2a,glm.2b,glm.2c, glm.3a, glm.3b, glm.3c,glm.3d,glm.3e) #90.27

summary(glm.3b)
 #                     Estimate Std. Error t value Pr(>|t|)    
#(Intercept)          4.5853393  0.5674501   8.081 1.63e-09 ***
#seasonSummer        -0.3559855  0.2307261  -1.543 0.131851    
#siteNR              -1.5511394  0.6311079  -2.458 0.019074 *  
#siteRI              -0.5065048  0.5954430  -0.851 0.400754    
#density_sqrt        -0.0004940  0.0001903  -2.595 0.013715 *  
#siteNR:density_sqrt  0.0009398  0.0002391   3.930 0.000382 ***
#siteRI:density_sqrt  0.0003175  0.0002124   1.495 0.143860    


#check for collinearity....
#library(mctest)
omcdiag(glm.3b) #chisq significant, indicates multicolinnearity
#check which variables collinear
imcdiag(glm.3b) # shows all except season collinear

#library(faraway)
vif(glm.3b)


tab_model(glm.3b, digits=4)

#cross validation
#library(boot)
cv.err <- cv.glm(df.div.coral, glm.3b)$delta #leave one out cross validation if you don't set K (I think)
cv.err.10 <- cv.glm(df.div.coral, glm.3b, K=10)$delta

muhat <- fitted(glm.3b)
coral.diag <- glm.diag(glm.3b)
(cv.err <- mean((glm.3b$y - muhat)^2/(1 - coral.diag$h)^2))

#trying with symbstate instead of density:
glm.0 <- glm(Shannon ~ 1, data=df.div.coral, family=gaussian())
glm.1a <- glm(Shannon ~ site, data=df.div.coral, family=gaussian())
glm.1b <- glm(Shannon ~ season, data=df.div.coral, family=gaussian())
glm.1c <- glm(Shannon ~ symbstate, data=df.div.coral, family=gaussian())

glm.2a <- glm(Shannon ~ season+site, data=df.div.coral, family=gaussian())
glm.2b <- glm(Shannon ~ season+symbstate, data=df.div.coral, family=gaussian())
glm.2c <- glm(Shannon ~ site+symbstate, data=df.div.coral, family=gaussian())

glm.3a <- glm(Shannon ~ season+site+symbstate, data=df.div.coral, family=gaussian())
glm.3b <- glm(Shannon ~ season+site*symbstate, data=df.div.coral, family=gaussian())
glm.3c <- glm(Shannon ~ season*site+symbstate, data=df.div.coral, family=gaussian())
glm.3d <- glm(Shannon ~ site+season*symbstate, data=df.div.coral, family=gaussian())
glm.3e <- glm(Shannon ~ season*site*symbstate, data=df.div.coral, family=gaussian())

AICc(glm.0,glm.1a,glm.1b,glm.1c, glm.2a,glm.2b,glm.2c, glm.3a, glm.3b, glm.3c,glm.3d,glm.3e)
#glm.1b best AIC (dif from raw)
summary(glm.1b)
#             Estimate Std. Error t value Pr(>|t|)    
#(Intercept)    3.9048     0.2001  19.511   <2e-16 ***
#seasonSummer  -0.6066     0.2451  -2.475   0.0177 *  

tab_model(glm.1b, digits=2)
```

```{r}
## GLM - Evenness ###

glm.0 <- glm(even ~ 1, data=df.div.coral, family=gaussian())
glm.1a <- glm(even ~ site, data=df.div.coral, family=gaussian())
glm.1b <- glm(even ~ season, data=df.div.coral, family=gaussian())
glm.1c <- glm(even ~ density_sqrt, data=df.div.coral, family=gaussian())

glm.2a <- glm(even ~ season+site, data=df.div.coral, family=gaussian())
glm.2b <- glm(even ~ season+density_sqrt, data=df.div.coral, family=gaussian())
glm.2c <- glm(even ~ site+density_sqrt, data=df.div.coral, family=gaussian())

glm.3a <- glm(even ~ season+site+density_sqrt, data=df.div.coral, family=gaussian())
glm.3b <- glm(even ~ season+site*density_sqrt, data=df.div.coral, family=gaussian())
glm.3c <- glm(even ~ season*site+density_sqrt, data=df.div.coral, family=gaussian())
glm.3d <- glm(even ~ site+season*density_sqrt, data=df.div.coral, family=gaussian())
glm.3e <- glm(even ~ season*site*density_sqrt, data=df.div.coral, family=gaussian())

AICc(glm.0,glm.1a,glm.1b,glm.1c, glm.2a,glm.2b,glm.2c, glm.3a, glm.3b, glm.3c,glm.3d,glm.3e)

# glm.1b best (dif from raw, was 1c) - lines up with effect of season above (for all types) too
summary(glm.1b) 
#             Estimate Std. Error t value Pr(>|t|)    
#(Intercept)    3.9048     0.2001  19.511   <2e-16 ***
#seasonSummer  -0.6066     0.2451  -2.475   0.0177 *  

tab_model(glm.1b, digits=2)

### Symbstate - evenness

glm.0 <- glm(even ~ 1, data=df.div.coral, family=gaussian())
glm.1a <- glm(even ~ site, data=df.div.coral, family=gaussian())
glm.1b <- glm(even ~ season, data=df.div.coral, family=gaussian())
glm.1c <- glm(even ~ symbstate, data=df.div.coral, family=gaussian())

glm.2a <- glm(even ~ season+site, data=df.div.coral, family=gaussian())
glm.2b <- glm(even ~ season+symbstate, data=df.div.coral, family=gaussian())
glm.2c <- glm(even ~ site+symbstate, data=df.div.coral, family=gaussian())

glm.3a <- glm(even ~ season+site+symbstate, data=df.div.coral, family=gaussian())
glm.3b <- glm(even ~ season+site*symbstate, data=df.div.coral, family=gaussian())
glm.3c <- glm(even ~ season*site+symbstate, data=df.div.coral, family=gaussian())
glm.3d <- glm(even ~ site+season*symbstate, data=df.div.coral, family=gaussian())
glm.3e <- glm(even ~ season*site*symbstate, data=df.div.coral, family=gaussian())

AICc(glm.0,glm.1a,glm.1b,glm.1c, glm.2a,glm.2b,glm.2c, glm.3a, glm.3b, glm.3c,glm.3d,glm.3e)
 #glm.3c best 

summary(glm.3c)
#                    Estimate Std. Error t value Pr(>|t|)    
#(Intercept)          0.82447    0.04706  17.521  < 2e-16 ***
#seasonSummer        -0.14554    0.03727  -3.905 0.000397 ***
#siteNR              -0.03489    0.06199  -0.563 0.577000    
#siteRI               0.04075    0.03831   1.064 0.294605    
#symbstateSymbiotic   0.02750    0.02538   1.083 0.285805    
#seasonSummer:siteNR  0.13067    0.06121   2.135 0.039649 *  
#seasonSummer:siteRI       NA         NA      NA       NA   

tab_model(glm.3c)

```


```{r GLM richness}
## GLM - richness ###
qqPlot(df.div.coral$Observed) # discrete, not continuous
shapiro.test(df.div.coral$Observed) #really sig/not normal
ggdensity(df.div.coral$Observed) #right skewed

#using poisson because counts; however, poisson is sensitive to assumption that mean is equal to variance; alternative is quasi-poisson or negative binomial
glm.0 <- glm(Observed ~ 1, data=df.div.coral, family=poisson())
glm.1 <- glm(Observed ~ season+site+density, data=df.div.coral, family=poisson())
#summary(glm.1)
glm.2 <- glm(Observed ~ season+site*density, data=df.div.coral, family=poisson())
#summary(glm.2)
glm.3 <- glm(Observed ~ season*site+density, data=df.div.coral, family=poisson())
#summary(glm.3)
glm.4 <- glm(Observed ~ site+season*density, data=df.div.coral, family=poisson())
#summary(glm.4)
glm.5 <- glm(Observed ~ season*site*density, data=df.div.coral, family=poisson())
#summary(glm.5)

AICc(glm.0,glm.1, glm.2, glm.3, glm.4, glm.5)
# glm.5 is best
summary(glm.5) 

#for glm.5:
#Coefficients: (2 not defined because of singularities)
#^^^why does this happen?! i think bc of CL...can I change CL to not be intercept?


# trying negative binomial
library(MASS)

glm.0 <- glm.nb(Observed ~ 1, data=df.div.coral)
glm.1a <- glm.nb(Observed ~ site, data=df.div.coral)
glm.1b <- glm.nb(Observed ~ season, data=df.div.coral)
glm.1c <- glm.nb(Observed ~ density_sqrt, data=df.div.coral)

glm.2a <- glm.nb(Observed ~ season+site, data=df.div.coral)
glm.2b <- glm.nb(Observed ~ season+density_sqrt, data=df.div.coral)
glm.2c <- glm.nb(Observed ~ site+density_sqrt, data=df.div.coral)

glm.3a <- glm.nb(Observed ~ season+site+density_sqrt, data=df.div.coral)
glm.3b <- glm.nb(Observed ~ season+site*density_sqrt, data=df.div.coral)
glm.3c <- glm.nb(Observed ~ season*site+density_sqrt, data=df.div.coral)
glm.3d <- glm.nb(Observed ~ site+season*density_sqrt, data=df.div.coral)
glm.3e <- glm.nb(Observed ~ season*site*density_sqrt, data=df.div.coral)

AICc(glm.0,glm.1a,glm.1b,glm.1c, glm.2a,glm.2b,glm.2c, glm.3a, glm.3b, glm.3c,glm.3d,glm.3e)

# glm.3b best (same as raw)

summary(glm.3b)
#                      Estimate Std. Error z value Pr(>|z|)    
#(Intercept)          5.4430490  0.4078756  13.345  < 2e-16 ***
#seasonSummer         0.0674280  0.1661036   0.406 0.684787    
#siteNR              -1.4592003  0.4547268  -3.209 0.001332 ** 
#siteRI              -0.5191887  0.4280676  -1.213 0.225181    
#density_sqrt        -0.0004707  0.0001384  -3.402 0.000669 ***
#siteNR:density_sqrt  0.0008273  0.0001732   4.776 1.79e-06 ***
#siteRI:density_sqrt  0.0002334  0.0001547   1.509 0.131291    

tab_model(glm.3b, digits=4)


### symbiotic state
glm.0 <- glm.nb(Observed ~ 1, data=df.div.coral)
glm.1a <- glm.nb(Observed ~ site, data=df.div.coral)
glm.1b <- glm.nb(Observed ~ season, data=df.div.coral)
glm.1c <- glm.nb(Observed ~ symbstate, data=df.div.coral)

glm.2a <- glm.nb(Observed ~ season+site, data=df.div.coral)
glm.2b <- glm.nb(Observed ~ season+symbstate, data=df.div.coral)
glm.2c <- glm.nb(Observed ~ site+symbstate, data=df.div.coral)

glm.3a <- glm.nb(Observed ~ season+site+symbstate, data=df.div.coral)
glm.3b <- glm.nb(Observed ~ season+site*symbstate, data=df.div.coral)
glm.3c <- glm.nb(Observed ~ season*site+symbstate, data=df.div.coral)
glm.3d <- glm.nb(Observed ~ site+season*symbstate, data=df.div.coral)
glm.3e <- glm.nb(Observed ~ season*site*symbstate, data=df.div.coral)

AICc(glm.0,glm.1a,glm.1b,glm.1c, glm.2a,glm.2b,glm.2c, glm.3a, glm.3b, glm.3c,glm.3d,glm.3e)
#null model best, glm.1a close second
summary(glm.1a)
#            Estimate Std. Error z value Pr(>|z|)    
#(Intercept)  4.44526    0.19374  22.945   <2e-16 ***
#siteNR       0.46054    0.24770   1.859    0.063 .  
#siteRI       0.07424    0.23510   0.316    0.752    

tab_model(glm.1a)
```

```{r GLM Chao1}

glm.0 <- glm(Chao1 ~ 1, data=df.div.coral, family=gaussian())
glm.1a <- glm(Chao1 ~ site, data=df.div.coral, family=gaussian())
glm.1b <- glm(Chao1 ~ season, data=df.div.coral, family=gaussian())
glm.1c <- glm(Chao1 ~ density_sqrt, data=df.div.coral, family=gaussian())

glm.2a <- glm(Chao1 ~ season+site, data=df.div.coral, family=gaussian())
glm.2b <- glm(Chao1 ~ season+density_sqrt, data=df.div.coral, family=gaussian())
glm.2c <- glm(Chao1 ~ site+density_sqrt, data=df.div.coral, family=gaussian())

glm.3a <- glm(Chao1 ~ season+site+density_sqrt, data=df.div.coral, family=gaussian())
glm.3b <- glm(Chao1 ~ season+site*density_sqrt, data=df.div.coral, family=gaussian())
glm.3c <- glm(Chao1 ~ season*site+density_sqrt, data=df.div.coral, family=gaussian())
glm.3d <- glm(Chao1 ~ site+season*density_sqrt, data=df.div.coral, family=gaussian())
glm.3e <- glm(Chao1 ~ season*site*density_sqrt, data=df.div.coral, family=gaussian())

AICc(glm.0,glm.1a,glm.1b,glm.1c, glm.2a,glm.2b,glm.2c, glm.3a, glm.3b, glm.3c,glm.3d,glm.3e)
#glm.3b best

summary(glm.3b)

#                     Estimate Std. Error t value Pr(>|t|)    
#(Intercept)          32.02923   43.05733   0.744 0.461918    
#seasonFall          -17.71509   27.60293  -0.642 0.525195    
#siteCL              161.88566   75.50262   2.144 0.039049 *  
#siteRI              123.88024   54.32057   2.281 0.028776 *  
#density_sqrt          0.05763    0.01732   3.327 0.002070 ** 
#siteCL:density_sqrt  -0.09835    0.02861  -3.438 0.001530 ** 
#siteRI:density_sqrt  -0.08109    0.02097  -3.867 0.000458 ***

effectsize::standardize_parameters(glm.3b, exp = TRUE)
#Parameter           | Std. Coef. |       95% CI
#-----------------------------------------------
#(Intercept)         |       1.66 | [1.04, 2.66]
#seasonFall          |       0.82 | [0.44, 1.51]
#siteCL              |       0.52 | [0.25, 1.08]
#siteRI              |       0.52 | [0.29, 0.95]
#density_sqrt        |       2.42 | [1.44, 4.06]
#siteCL:density_sqrt |       0.22 | [0.09, 0.52]
#siteRI:density_sqrt |       0.29 | [0.15, 0.54]

tab_model(glm.3b)
```

```{r GLM InvSimpson}
## GLM - InvSimpson ###
#Chao1 is estimator based on abundance (estimates # of species in a community)

ggplot(df.div.coral, aes(x=InvSimpson))+
  geom_histogram(fill="#69b3a2", color="#e9ecef", alpha=0.9)
shapiro.test(df.div.coral$InvSimpson) #very non-normal
#log transform
#df.div.coral.log <- df.div.coral
df.div.coral.log$InvSimpson <- log(df.div.coral.log$InvSimpson)
ggplot(df.div.coral.log, aes(x=InvSimpson))+
  geom_histogram(fill="#69b3a2", color="#e9ecef", alpha=0.9) 
shapiro.test(df.div.coral.log$InvSimpson)
# log-transformed seems better

#models:
glm.1 <- glm(InvSimpson ~ season+site+density, data=df.div.coral.log, family=Gamma())
#summary(glm.1)
glm.2 <- glm(InvSimpson ~ season+site+density+site*density, data=df.div.coral.log, family=Gamma())
#summary(glm.2)
glm.3 <- glm(InvSimpson ~ season+site+density+season*site, data=df.div.coral.log, family=Gamma())
#summary(glm.3)
glm.4 <- glm(InvSimpson ~ season+site+density+season*density, data=df.div.coral.log, family=Gamma())
#summary(glm.4)
glm.5 <- glm(InvSimpson ~ season+site+density+season*site*density, data=df.div.coral.log, family=Gamma())
#summary(glm.5)

AICc(glm.1, glm.2, glm.3, glm.4, glm.5)
# glm.2 best AIC
summary(glm.2)
#                Estimate Std. Error t value Pr(>|t|)   
#(Intercept)     0.561777   0.261849   2.145  0.03894 * 
#seasonSummer    0.199690   0.129506   1.542  0.13208   
#siteNR          0.431138   0.286998   1.502  0.14201   
#siteRI          0.289421   0.267720   1.081  0.28707   
#density         0.019753   0.006431   3.072  0.00410 **
#siteNR:density -0.024227   0.007065  -3.429  0.00157 **
#siteRI:density -0.019107   0.006653  -2.872  0.00689 **
```

```{r GLM excluding CL - sanity check?}
df.div.coral.NR.RI$season <- as.factor(df.div.coral.NR.RI$season)
df.div.coral.NR.RI$site <- as.factor(df.div.coral.NR.RI$site)
df.div.coral.NR.RI$density <- as.numeric(df.div.coral.NR.RI$density)
str(df.div.coral.NR.RI)


glm.1 <- glm(Shannon ~ season+site+density, data=df.div.coral.NR.RI, family=Gamma())
#summary(glm.1)
glm.2 <- glm(Shannon ~ season+site+density+site*density, data=df.div.coral.NR.RI, family=Gamma())
#summary(glm.2)
glm.3 <- glm(Shannon ~ season+site+density+season*site, data=df.div.coral.NR.RI, family=Gamma())
#summary(glm.3)
glm.4 <- glm(Shannon ~ season+site+density+season*density, data=df.div.coral.NR.RI, family=Gamma())
#summary(glm.4)
glm.5 <- glm(Shannon ~ season+site+density+season*site*density, data=df.div.coral.NR.RI, family=Gamma())
#summary(glm.5)

AICc(glm.1, glm.2, glm.3, glm.4, glm.5)
#glm.2 best model (still! woo!)

summary(glm.2)

```



## Playing with microshades
```{r microshades setup}
#remotes::install_github("KarstensLab/microshades")
library("microshades")
#remotes::install_github("mikemc/speedyseq")
library("speedyseq")

#all
ps_glom <- tax_glom(ps.rare, "Genus")
ps0 <- transform_sample_counts(ps_glom, function(x) x / sum(x))
ps1 <- merge_samples(ps0, "fullname")
ps.all <- transform_sample_counts(ps1, function(x) x / sum(x))

# include other variables in ps.all (beyond just fullname)
seqtab.cleanest <- read.csv("oculina16s_seqtab.cleanest.csv",row.names=1)
ps.cleanest <- phyloseq(otu_table(seqtab.cleanest, taxa_are_rows=FALSE), 
                    sample_data(samdf), 
                    tax_table(taxa2))
ps.cleanest 
#save(ps.cleanest,file="ps.cleanest.Rdata")
samdf.cleanest <- data.frame(ps.cleanest@sam_data)

#for rarefied data, remove everything < 2000
row.names.remove <- c("D6","F3","H2","F7","H4","H7")
samdf.rare <- samdf.cleanest[!(row.names(samdf.cleanest) %in% row.names.remove), ]

#variable1 = as.character(get_variable(samdf.rare, "type"))
#sample_data(ps.all)$type <- mapply(paste0, variable1)
all(rownames(samdf.rare) %in% sample_names(ps.all))
samdf.rare = sample_data(samdf.rare)
merge_phyloseq(ps.all, samdf.rare)
head(sample_data(ps.all))

#overall by all_id
ps_glom.rare <- tax_glom(ps.rare, "Genus")
ps0.rare <- transform_sample_counts(ps_glom.rare, function(x) x / sum(x))
ps1.rare <- merge_samples(ps0.rare, "all_id")
ps.rare.allid <- transform_sample_counts(ps1.rare, function(x) x / sum(x))

# overall by type
ps1.rare.type <- merge_samples(ps0.rare, "type")
ps.rare.type <- transform_sample_counts(ps1.rare.type, function(x) x / sum(x))

# overall coral by all_id
ps.coral <- subset_samples(ps.rare,type=="coral")
ps_glom.coral <- tax_glom(ps.coral, "Genus")
ps0.coral <- transform_sample_counts(ps_glom.coral, function(x) x / sum(x))
ps1.coral <- merge_samples(ps0.coral, "all_id")
ps.coral.allid <- transform_sample_counts(ps1.coral, function(x) x / sum(x))

# overall coral by site
ps1.coral.site <- merge_samples(ps.coral, "site")
ps.coral.site <- transform_sample_counts(ps1.coral.site, function(x) x / sum(x))

# overall coral by season
ps1.coral.season <- merge_samples(ps.coral, "season")
ps.coral.season <- transform_sample_counts(ps1.coral.season, function(x) x / sum(x))

# overall coral by symbstate
ps1.coral.symbstate <- merge_samples(ps.coral, "symbstate")
ps.coral.symbstate <- transform_sample_counts(ps1.coral.symbstate, function(x) x / sum(x))

# Use microshades function prep_mdf to agglomerate, normalize, and melt the phyloseq object
mdf.prep.rare <- prep_mdf(ps.rare, subgroup_level="Genus")
mdf.prep <- prep_mdf(ps.all, subgroup_level="Genus")
mdf.prep.rare.allid <- prep_mdf(ps.rare.allid, subgroup_level = "Genus")
mdf.prep.rare.type <- prep_mdf(ps.rare.type, subgroup_level = "Family")
mdf.prep.allid <- prep_mdf(ps.coral.allid, subgroup_level = "Family")
mdf.prep.site <- prep_mdf(ps.coral.site, subgroup_level = "Family")
mdf.prep.season <- prep_mdf(ps.coral.season, subgroup_level = "Family")
mdf.prep.symbstate <- prep_mdf(ps.coral.symbstate, subgroup_level = "Family")

# Create a color object for the specified data
color.obj.rare <- create_color_dfs(mdf.prep.rare, group_level = "Phylum", subgroup_level = "Genus", cvd = TRUE)
color.obj <- create_color_dfs(mdf.prep, group_level = "Phylum", subgroup_level = "Genus", cvd = TRUE)
color.obj.rare.allid <- create_color_dfs(mdf.prep.rare.allid, group_level = "Phylum", subgroup_level = "Family", cvd = TRUE)
color.obj.rare.type <- create_color_dfs(mdf.prep.rare.type, group_level = "Phylum", subgroup_level = "Family", cvd = TRUE)
color.obj.allid <- create_color_dfs(mdf.prep, group_level = "Phylum", subgroup_level = "Family", cvd = TRUE)
color.obj.site <- create_color_dfs(mdf.prep.site, group_level = "Phylum", subgroup_level = "Family", cvd = TRUE)
color.obj.season <- create_color_dfs(mdf.prep.season, group_level = "Phylum", subgroup_level = "Family", cvd = TRUE)
color.obj.symbstate <- create_color_dfs(mdf.prep.symbstate, group_level = "Phylum", subgroup_level = "Family", cvd = TRUE)


# Extract
mdf.rare <- color.obj.rare$mdf
cdf.rare <- color.obj.rare$cdf
mdf <- color.obj$mdf
cdf <- color.obj$cdf
mdf.rare.allid <- color.obj.rare$mdf #contains sample data and abundance info
cdf.rare.allid <- color.obj.rare$cdf #contains color mapping info
mdf.rare.type <- color.obj.rare.type$mdf 
cdf.rare.type <- color.obj.rare.type$cdf 
mdf.coral.allid <- color.obj$mdf.allid
cdf.coral.allid <- color.obj$cdf.allid
mdf.coral.site <- color.obj.site$mdf
cdf.coral.site <- color.obj.site$cdf
mdf.coral.season <- color.obj.season$mdf
cdf.coral.season <- color.obj.season$cdf
mdf.coral.symbstate <- color.obj.symbstate$mdf
cdf.coral.symbstate <- color.obj.symbstate$cdf
```


```{r microshades initial plots}
plot1 <- plot_microshades(mdf, cdf, group_label = "Phylum Genus")
plot1.all <- plot_microshades(mdf.rare, cdf.rare, group_label = "Phylum Family")
plot1.all.type <- plot_microshades(mdf.rare.type, cdf.rare.type, group_label = "Phylum Family")
plot1.site <- plot_microshades(mdf.coral.site, cdf.coral.site, group_label = "Phylum Family")
plot1.season <- plot_microshades(mdf.coral.season, cdf.coral.season, group_label = "Phylum Family")
plot1.symbstate <- plot_microshades(mdf.coral.symbstate, cdf.coral.symbstate, group_label = "Phylum Family")

```

```{r all samples reordered by group}
# reorder_samples_by will change the order of samples based on an abundance of a specified subgroup taxonomy
new.sample.order <- reorder_samples_by(mdf, cdf, order = "Thiovulum", group_level = "Phylum", subgroup_level = "Genus", sink_abundant_groups = TRUE)

mdf.new.sample.order <-new.sample.order$mdf
cdf.new.sample.order <-new.sample.order$cdf

plot_2 <- plot_microshades(mdf.new.sample.order, cdf.new.sample.order, group_label = "Phylum Family")

plot_2 + scale_y_continuous(labels = scales::percent, expand = expansion(0)) +
  theme(legend.key.size = unit(0.2, "cm"), text=element_text(size=10)) +
  theme(axis.text.x = element_text(size= 6))
```

```{r all_id reordered by group}
# reorder_samples_by will change the order of samples based on an abundance of a specified subgroup taxonomy
new.sample.order.coral <- reorder_samples_by(mdf.rare, cdf.rare, order = "Cyanobium_PCC-6307", group_level = "Phylum", subgroup_level = "Genus", sink_abundant_groups = TRUE)

mdf.new.sample.order <-new.sample.order.coral$mdf
cdf.new.sample.order <-new.sample.order.coral$cdf

plot_2 <- plot_microshades(mdf.new.sample.order, cdf.new.sample.order, group_label = "Phylum Genus")

plot_2 + scale_y_continuous(labels = scales::percent, expand = expansion(0)) +
  theme(legend.key.size = unit(0.2, "cm"), text=element_text(size=10)) +
  theme(axis.text.x = element_text(size= 6))
```

```{r reorder samples by abundance of group levels only}
# reorder_samples_by will change the order of samples based on an abundance of a specified subgroup taxonomy
# The default subgroup_level is "Genus" 
new.group.order <- reorder_samples_by(mdf.rare, cdf.rare, group_level = "Phylum", subgroup_level = "Family", sink_abundant_groups = TRUE)

mdf.new.group.order <-new.group.order$mdf
cdf.new.group.order <-new.group.order$cdf

plot_3 <- plot_microshades(mdf.new.group.order, cdf.new.group.order, group_label = "Phylum Family")

plot_3 + scale_y_continuous(labels = scales::percent, expand = expansion(0)) +
  theme(legend.key.size = unit(0.2, "cm"), text=element_text(size=10)) +
  theme(axis.text.x = element_text(size= 6))
```
```{r}
# reorder_samples_by will change the order of samples based on Family an abundance of a group and reorder the Phylum to sink the most abundant groups
reordered <- reorder_samples_by(mdf.rare, cdf.rare, order = "Other", group_level = "Phylum", subgroup_level = "Family",sink_abundant_groups=TRUE)

mdf.reordered <-reordered$mdf
cdf.reordered <-reordered$cdf

plot_3 <- plot_microshades(mdf.reordered, cdf.reordered, group_label = "Phylum Family")

plot_3 + scale_y_continuous(labels = scales::percent, expand = expansion(0)) +
  theme(legend.key.size = unit(0.2, "cm"), text=element_text(size=10)) +
  theme(axis.text.x = element_text(size= 6))
```


```{r facet wrap}
# remove current Sample column (currently a repeat of id), and then rename fullname = Sample
#library(dplyr)
mdf.rare <- mdf.rare %>%
  select(-Sample) %>%
  rename(Sample = fullname) %>%
  rename(number = sample)

plot.rare <- plot_microshades(mdf.rare, cdf.rare, group_label = "Phylum Genus")

plot.rare + 
  theme(legend.key.size = unit(0.5, "cm"), text=element_text(size=10)) +
  theme(axis.text.x = element_text(size= 6)) +
  facet_grid(~type, scales = "free_x", space="free") +
  theme (strip.text.x = element_text(size = 10))
  
```



# So... I think I have an issue with multicollinearity in my data, which could make some variables significant predictors when they shouldn't be. Going to try lasso regression as that can apparently help.
```{r lasso caret}
str(df.div.coral)
df.div.coral.sha <- df.div.coral %>% dplyr::select(Shannon, site, season, density_sqrt)
str(df.div.coral.sha)

library(caret)
model <- train(
  Shannon ~ .,
  data = df.div.coral.sha,
  method = 'lasso'
)
model
plot(model)
plot(varImp(model))

#pre-process data
model2 <- train(
  Shannon ~ .,
  data = df.div.coral.sha,
  method = 'lasso',
  preProcess = c("center", "scale") #center and scale the data
)
model2

#training
set.seed(1)

inTraining <- createDataPartition(df.div.coral.sha$Shannon, p = .80, list = FALSE)
training <- df.div.coral.sha[inTraining,]
testing  <- df.div.coral.sha[-inTraining,]

model3 <- train(
  Shannon ~.,
  data = training,
  method = 'lasso',
  preProcess = c("center", "scale")
)
model3

test.features = subset(testing, select=-c(Shannon))
test.target = subset(testing, select=Shannon)[,1]

predictions = predict(model3, newdata = test.features)

# RMSE
sqrt(mean((test.target - predictions)^2))
#0.816311


#cross validation
ctrl <- trainControl(
  method = "cv",
  number = 10,
)


model4 <- train(
  Shannon ~ .,
  data = training,
  method = 'lasso',
  preProcess = c("center", "scale"),
  trControl = ctrl
)
model4

test.features = subset(testing, select=-c(Shannon))
test.target = subset(testing, select=Shannon)[,1]

predictions = predict(model4, newdata = test.features)

# RMSE
sqrt(mean((test.target - predictions)^2))
#0.7229497


#R2
cor(test.target, predictions) ^ 2
#0.4864439


tuneGrid <- expand.grid(
  .fraction = seq(0, 1, by = 0.1)
)


model5 <- train(
  Shannon ~ .,
  data = training,
  method = 'lasso',
  preProcess = c("center", "scale"),
  trControl = ctrl,
  tuneGrid = tuneGrid
)

model5
plot(model5)
plot(varImp(model5))

```

# Trying lasso regression with glmnet package
```{r}
#library(glmnet)
#define response variable
y <- df.div.coral.sha$Shannon

#define matrix of predictor variables
x <-data.matrix(df.div.coral.sha %>% dplyr::select(site:density_sqrt))

#perform k-fold cross-validation to find optimal lambda value
cv_model <- cv.glmnet(x, y, alpha = 1)

#find optimal lambda value that minimizes test MSE
best_lambda <- cv_model$lambda.min
best_lambda

#[1] 0.03693455

#produce plot of test MSE by lambda value
plot(cv_model) 

#find coefficients of best model
best_model <- glmnet(x, y, alpha = 1, lambda = best_lambda)
coef(best_model)
#4 x 1 sparse Matrix of class "dgCMatrix"
#                        s0
#(Intercept)   3.115196e+00
#site         -1.308859e-01
#season        5.675169e-01
#density_sqrt -4.229753e-05

#use fitted best model to make predictions
y_predicted <- predict(best_model, s = best_lambda, newx = x)

#find SST and SSE
sst <- sum((y - mean(y))^2)
sse <- sum((y_predicted - y)^2)

#find R-Squared
rsq <- 1 - sse/sst
rsq

#[1] 0.1782721 so small!

```


### Faith's D


## Phylogenetic diversity (Faith's D)

Tutorial from dada2 author [here](https://f1000research.com/articles/5-1492/v2)

```{r packages phylo d, eval=FALSE}
#install.packages('devtools')
#library(devtools)
#devtools::install_github('twbattaglia/btools')
#library(btools)
```


(I'm not running the following chunk every time, because only need to generate the file once)

```{r fasta file generation, eval=FALSE}
raw.otu  <- as.matrix(ps.rare@otu_table)
raw.taxa <- data.frame(ps.rare@tax_table)
rownames(raw.taxa)==colnames(raw.otu)
 
colnames(raw.otu) <- raw.taxa$V8
ids <- rownames(raw.taxa)

path="~/oculina/diversity/analyze_asv_table/oculina16s_rev.less.fasta"
uniquesToFasta(raw.otu, path, ids = ids, mode = "w", width = 20000)
```


Actual analysis part: 

(I'm not running the following chunk every time, because only need to generate the files once)

```{r phylo d, eval=FALSE}
seqs <- getSequences("oculina16s_rev.less.fasta")
names(seqs) <- seqs # This propagates to the tip labels of the tree
saveRDS(seqs,file="phylo.seqs.less.rds")
#also doing the same thing with a .fasta file post-trimming to see if it makes a difference:
#seqs <- getSequences("mr16s_rev.cleanest.trimmed copy.fasta")
#names(seqs) <- seqs # This propagates to the tip labels of the tree
#saveRDS(seqs,file="phylo.seqs.rev.cleanest.trimmed.rds")
```


Doing this next part in the cluster because it takes forever

```{bash terminal phylo things, eval=FALSE}
##script phylo.R looks like this now:
# library(dada2)
# library(phangorn)
# library(DECIPHER)
#
# seqs <- readRDS("./phylo.seqs.rev.cleanest.rds")
# alignment <- AlignSeqs(DNAStringSet(seqs), anchor=NA)
# phang.align <- phyDat(as(alignment, "matrix"), type="DNA")
# dm <- dist.ml(phang.align)
# treeNJ <- NJ(dm) # Note, tip order != sequence order
# fit = pml(treeNJ, data=phang.align)
#
# ## negative edges length changed to 0!
# 
# fitGTR <- update(fit, k=4, inv=0.2)
# fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
#                     rearrangement = "stochastic", control = pml.control(trace = 0))
# detach("package:phangorn", unload=TRUE)
# saveRDS(fitGTR, file="./phylo.fitgtr.rev.cleanest.rds")

### In terminal:
#nano phylo.sh #to create and edit text file
#phylo.sh looks like this: 
  # #!/bin/bash
  # #SBATCH -p general
  # #SBATCH -t 1-
  # #SBATCH -N 1
  # #SBATCH -n 1
  # #SBATCH --mem=8gb
  # #SBATCH --output <phylo.fitgtr.less.rds>
  
  # module add r
  # module add rstudio
  
  # Rscript phylo.R

##saved output as: phylo.fitgtr.less.rds
```

Back in R
```{r}
#BiocManager::install("DESeq2")
#BiocManager::install("genefilter")
library(DESeq2)
library(genefilter)
#devtools::install_github('twbattaglia/btools')
library(btools)
setwd("~/oculina/diversity/analyze_asv_table")
fitGTR <- readRDS("phylo.fitgtr.less.rds")

#new phyloseq object:
taxa.less <- data.frame(ps.less@tax_table)
seqtab.less <- data.frame(ps.less@otu_table)
taxa.less$sqs <- row.names(taxa.less) 
taxa.less$sqs == colnames(seqtab.less)
row.names(taxa.less) <- taxa.less$V8
colnames(seqtab.less) <- taxa.less$V8
row.names(taxa.less) == colnames(seqtab.less)
taxa.less <- as.matrix(taxa.less)
ps.less.tree <- phyloseq(otu_table(seqtab.less, taxa_are_rows = FALSE),
                         sample_data(samdf),
                         tax_table(taxa.less),
                         phy_tree(fitGTR$tree))

pd.div <- estimate_pd(ps.less.tree)
row.names(df.div) <- df.div$id
df.div.pd <- merge(df.div,pd.div,by=0)

#saveRDS(ps.less.tree,file="ps.less.tree.Rdata")
## saving diversity data frame ##
#save & read back in as needed
#write.csv(df.div.pd,file="oculina16s_diversity_less.csv") #saving
#df.div <- read.csv("oculina16s_diversity_less.csv",row.names=1,header=TRUE) #reading back in

```

