---
title: "oculina16s_analysis_raw"
output:
  word_document: default
  html_document: default
date: '2022-08-17'
editor_options: 
  chunk_output_type: console
---
### Based on DADA2 Pipeline 1.16 Walkthrough & Nicola Kriefall's Moorea holobiont analysis

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup

## Packages
```{r packages}
library(ggplot2)
library(cowplot)
library(phyloseq)
library(car)
library(ggpubr)
library(vegan)
library(dada2)
library(stats)
library(MuMIn)
library(sjPlot)
library(webshot)
library(MASS)
library(btools)
library(gtsummary)
```


## Read in data

```{r read in data}
#setwd("~/oculina/data/")
samdf <- read.csv("~/oculina/data/oculina16s_sampledata_symdens.csv",header=TRUE) 
row.names(samdf) <- samdf$id
load("~/oculina/data/taxa2.rev.Rdata")
```

## Phyloseq objects

```{r making phyloseq objects, eval=FALSE}
seqtab.less <- read.csv("~/oculina/data/oculina16s_rev_seqtab.less",row.names=1)
load("~/oculina/data/ps.less_rev.Rdata")
ps.less #7005 taxa, 71 samples
```

```{r}
#tried to do merge_phyloseq so that I could easily add new metadata but can't get it to work
#samdf.new <- read.csv("~/oculina/data/oculina16s_sampledata_symdens.csv")
#density.new <- samdf.new %>% select(id, density)
#samdf.sd <- density.new %>% rename(density.new = density)
#rownames(samdf.sd) <- samdf.new$id
#row.names.remove <- c("D6","H2","G9","F10","G10","H10")
#samdf.sd <- samdf.sd[!(row.names(samdf.sd) %in% row.names.remove), ]

#ps.less.new <- merge_phyloseq(ps.less, samdf.sd)
#head(sample_data(ps.less.new))
#samdf.less <- sample_data(ps.less.new)

#install.packages("arsenal")
#library("arsenal")

#summary(comparedf(samdf.sd, samdf.less))

```


# Diversity

[Notes from phyloseq author](https://rdrr.io/bioc/phyloseq/man/estimate_richness.html)
Visualize alpha-diversity - Should be done on raw, untrimmed dataset

```{r generate div metrics}
df <- data.frame(estimate_richness(ps.less, split=TRUE, measures=c("Shannon","Simpson", "InvSimpson","Observed")))
# NOTE: I get a warning that my data doesn't contain singletons, but this is untrimmed raw data, so I'm not sure why this is the case. I do not get this error when I use ps.rare, presumably because rarefying data caused some data to become singletons.
# NOTE from DADA2 creator: DADA2 does not call singletons, due to the difficulty of differentiating rare singleton errors from real singleton variants. This means you should not use the output of DADA2 to estimate richness (eg. Chao S1). However, you shouldn't have been doing that with the output of other methods either, as the high levels of FP singletons made richness estimates wrong anyway. Right now, I don't think a method exists that can make valid richness estimates from high-throughput amplicon data due to the difficulty of calling singletons accurately, and the sensitivity of richness estimation to the number of singletons. Other measures of diversity that aren't totally reliant on singletons, eg. Shannon/Simpson, are valid to use, and you can ignore the warning in phyloseq when calculating those measures.
# CONCLUSION: do not use Chao1 in analysis; Shannon and Simpson should be okay

df$id <- rownames(df)
df.div <- merge(df,samdf,by="id") #add sample data

#transform density
#library(e1071)
df.div.coral <- subset(df.div, type=="coral")
skewness(df.div.coral$density) #very right skewed (1.93)
df.div.coral$density_sqrt <- sqrt(df.div.coral$density)

#df.div$density <- as.numeric(df.div$density)
#df.div <- df.div %>% mutate(density_log = log(density))
#shannon diversity divided by species richness
df.div$even <- df.div$Shannon/(log(df.div$Observed))
df.div.NR <- subset(df.div,site=="NR")
df.div.RI <- subset(df.div,site=="RI")
df.div.CL <- subset(df.div,site=="CL")
df.div.summer <- subset(df.div,season=="Summer")
df.div.fall <- subset(df.div,season=="Fall")
df.div.ss <- subset(df.div, type=="sediment" | type =="seawater")
df.div.coral.NR <- subset(df.div.coral,site=="NR")
df.div.coral.RI <- subset(df.div.coral,site=="RI")
df.div.coral.CL <- subset(df.div.coral,site=="CL")
df.div.coral.NR.RI <- subset(df.div.coral,site %in% c("NR","RI"))

load("~/oculina/data/ps.less.tree.RDS")
pd.div <- estimate_pd(ps.less.tree)
row.names(df.div) <- df.div$id
df.div.pd <- merge(df.div,pd.div,by=0)
```

```{r Summary statistics all types}
library(gtsummary)

# all types
df1 <- df.div %>% select(Shannon, even, Observed, Chao1, type)
tbl.all.site <- tbl_summary(df1, by = type) %>%
  add_p(test = Shannon ~ "aov") %>% # test for a difference between groups
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()

aov.sha <- (aov(Shannon ~ type, data=df1))
TukeyHSD(aov.sha)
dunnTest(even~type, data=df1, method="bonferroni") #coral-sed, sw-sed sig
dunnTest(Observed~type, data=df1, method="bonferroni") #all sig
dunnTest(Chao1~type, data=df1, method="bonferroni") #coral-sed sig

df2 <- df.div %>% select(Shannon, even, Observed, Chao1, season)
tbl_summary(df2, by = season) %>%
  add_p(test= everything() ~ "kruskal.test") %>% # test for a difference between groups
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()

df.sha <- df2 %>% select(Shannon, season)
tbl_summary(df.sha, by=season) %>%
  add_p(test = Shannon ~ "aov") %>%
  bold_labels()

df2 <- df.div %>% select(Shannon, even, Observed, Chao1, site)
tbl_summary(df2, by = site) %>%
  add_p(test= everything() ~ "kruskal.test") %>% # test for a difference between groups
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()
df.sha <- df2 %>% select(Shannon, site)
tbl_summary(df.sha, by=site) %>%
  add_p(test = Shannon ~ "aov") %>%
  bold_labels()


# Faith's D
df <- df.div.pd %>% select(PD, type)
tbl_summary(df, by = type) %>%
  add_p() %>% # test for a difference between groups
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()


# look at season differences for sed + sw
##sw
df.sw.seas <- df.div %>%
  subset(type == "seawater") %>%
  select(Shannon, even, Observed, Chao1, season)

tbl_summary(df.sw.seas, by = season) %>%
  add_p(test= everything() ~ "kruskal.test") %>%
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()

df.sw.seas.sha <- df.sw.seas %>% select(Shannon, season)
tbl_summary(df.sw.seas.sha, by=season) %>%
  add_p(test = Shannon ~ "aov") %>%
  bold_labels()

# Faith's D for sw seasons
df <- df.div.pd %>% subset(type == "seawater") %>% select(PD, season)
tbl_summary(df, by = season) %>%
  add_p(test= everything() ~ "kruskal.test") %>% # test for a difference between groups
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()


df.sw.site <- df.div %>%
  subset(type == "seawater") %>%
  select(Shannon, even, Observed, Chao1, site)

tbl_summary(df.sw.site, by = site) %>%
  add_p(test= everything() ~ "kruskal.test") %>%
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()

df.sw.site.sha <- df.sw.site %>% select(Shannon, site)
tbl_summary(df.sw.site.sha, by=site) %>%
  add_p(test = Shannon ~ "aov") %>%
  bold_labels()
# Faith's D for sw sites
df <- df.div.pd %>% subset(type == "seawater") %>% select(PD, site)
tbl_summary(df, by = site) %>%
  add_p() %>% # test for a difference between groups
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()

## sediment
df.sed.seas <- df.div %>%
  subset(type == "sediment") %>%
  select(Shannon, even, Observed, Chao1, season)

tbl_summary(df.sed.seas, by = season) %>%
  add_p(test= everything() ~ "kruskal.test") %>%
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()

df.sed.seas.sha <- df.sed.seas %>% select(Shannon, season)
tbl_summary(df.sed.seas.sha, by=season) %>%
  add_p(test = Shannon ~ "aov") %>%
  bold_labels()

# Faith's D for sw seasons
df <- df.div.pd %>% subset(type == "sediment") %>% select(PD, season)
tbl_summary(df, by = season) %>%
  add_p(test= everything() ~ "kruskal.test") %>% # test for a difference between groups
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()

df.sed.site <- df.div %>%
  subset(type == "sediment") %>%
  select(Shannon, even, Observed, Chao1, site)

tbl_summary(df.sed.site, by = site) %>%
  add_p(test= everything() ~ "kruskal.test") %>%
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()

df.sed.site.sha <- df.sed.site %>% select(Shannon, site)
tbl_summary(df.sed.site.sha, by=site) %>%
  add_p(test = Shannon ~ "aov") %>%
  bold_labels()

# Faith's D for sw sites
df <- df.div.pd %>% subset(type == "sediment") %>% select(PD, site)
tbl_summary(df, by = site) %>%
  add_p(test= everything() ~ "kruskal.test") %>% # test for a difference between groups
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()

tbl.all.site %>%
  as_flex_table() %>%
  save_as_image(path ="~/oculina/diversity/figures/tbl.all.site.png") # use extensions .html .tex .ltx .rtf

```

```{r summary statistics coral}
df1 <- df.div.coral %>% select(Shannon, even, Observed, Chao1, site)
tbl.coral.seas <- tbl_summary(df1, by = site) %>%
  add_p(test= everything() ~ "kruskal.test") %>% # test for a difference between groups
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()
df.coral.site.sha <- df.div.coral %>% select(Shannon, site)
tbl_summary(df.coral.site.sha, by=site) %>%
  add_p(test = Shannon ~ "aov") %>%
  bold_labels()

aov.sha <- (aov(Shannon ~ site, data=df1))
TukeyHSD(aov.sha)

#post hoc test for kruskall wallis is the dunn test
#library(FSA)
dunnTest(even~site, data=df1, method="bonferroni") #CL-RI sig dif, others not

df2 <- df.div.coral %>% select(Shannon, even, Observed, Chao1, season)
tbl.coral.seas <- tbl_summary(df2, by = season) %>%
  add_p(test= everything() ~ "kruskal.test") %>% # test for a difference between groups
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()
df.coral.seas.sha <- df.div.coral %>% select(Shannon, season)
tbl_summary(df.coral.seas.sha, by=season) %>%
  add_p(test = Shannon ~ "aov") %>%
  bold_labels()


df3 <- df.div.coral %>% select(Shannon, even, Observed, Chao1, symbstate)
tbl.coral.ss <- tbl_summary(df3, by = symbstate) %>%
  add_p(test= everything() ~ "kruskal.test") %>% # test for a difference between groups
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()
df.coral.ss.sha <- df.div.coral %>% select(Shannon, symbstate)
tbl_summary(df.coral.ss.sha, by=symbstate) %>%
  add_p(test = Shannon ~ "aov") %>%
  bold_labels()

# check for interaction
df1 <- df.div.coral %>% select(Shannon, even, Observed, Chao1, site, density_sqrt, season)
lm <- (lm(data=df1, Shannon ~ site * density_sqrt)) #intxn sig (***)
tab_model(lm, digits = 4)
lm <- (lm(data=df1, even ~ site * density_sqrt)) #intxn sig (*)
summary(lm(data=df1, Observed ~ site * density_sqrt)) #intxn sig (***)
summary(lm(data=df1, Chao1 ~ site * density_sqrt)) #intxn ***

df1 <- df.div.coral %>% select(Shannon, even, Observed, Chao1, site, symbstate, season)
lm <- (lm(data=df1, Shannon ~ site * symbstate)) #intxn sig (***)
summary(lm) #symb state significant, and intxn sig (for RI)

# Faith's D
df <- df.div.coral.pd %>% select(PD, symbstate)
tbl.coral.ss <- tbl_summary(df, by = symbstate) %>%
  add_p(test= everything() ~ "kruskal.test") %>% # test for a difference between groups
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()

df <- df.div.coral.pd %>% select(PD, season)
tbl_summary(df, by = season) %>%
  add_p(test= everything() ~ "kruskal.test") %>% # test for a difference between groups
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()

df <- df.div.coral.pd %>% select(PD, site)
tbl_summary(df, by = site) %>%
  add_p(test= everything() ~ "kruskal.test") %>% # test for a difference between groups
  modify_header(label = "***Alpha Diversity***") %>% # update the column header
  bold_labels()


```



### GLMS - Alpha Diversity Indices

```{r compare types (coral, sed, sw)}
df.div$season <- as.factor(df.div$season)
df.div$site <- as.factor(df.div$site)
df.div$type <- as.factor(df.div$type)
df.div$density <- as.numeric(df.div$density)

str(df.div)

glm.0 <- glm(Shannon ~ 1, data=df.div, family=gaussian())
glm.1 <- glm(Shannon ~ type, data=df.div, family=gaussian)
glm.2 <- glm(Shannon ~ type + season + site, data=df.div, family=gaussian)
glm.3 <- glm(Shannon ~ type + season * site, data=df.div, family=gaussian)
glm.4 <- glm(Shannon ~ type * season + site, data=df.div, family=gaussian)
glm.5 <- glm(Shannon ~ type * season * site, data=df.div, family=gaussian)

AICc(glm.0, glm.1, glm.2, glm.3, glm.4, glm.5)
#glm.1 is best model
summary(glm.1)
#             Estimate Std. Error t value Pr(>|t|)    
#(Intercept)    3.4760     0.1223  28.424  < 2e-16 ***
#typeseawater   0.3132     0.2583   1.213    0.229    
#typesediment   1.6119     0.2583   6.240 3.22e-08 ***

#calc diff in chi square statistics for null model and the model with independent variances (deviances) and dif in df
1-pchisq(71.991-45.761, 70-68)
#2.014781e-06 
#reject null hypothesis that there is no different b/w null and model with variables

tab_model(glm.1) #makes a nice looking table
tab_model(glm.1, file="./stats/sha.types.glm.gauss.html")

webshot("./stats/sha.types.glm.gauss.html", "./stats/sha.types.glm.gauss.png")


### Evenness - all types

glm.0 <- glm(even ~ 1, data=df.div, family=gaussian())
glm.1 <- glm(even ~ type, data=df.div, family=gaussian)
glm.2 <- glm(even ~ type + season + site, data=df.div, family=gaussian)
glm.3 <- glm(even ~ type + season * site, data=df.div, family=gaussian)
glm.4 <- glm(even ~ type * season + site, data=df.div, family=gaussian)
glm.5 <- glm(even ~ type * season * site, data=df.div, family=gaussian)

AICc(glm.0, glm.1, glm.2, glm.3, glm.4, glm.5)
# glm.4 best (-153.0)
summary(glm.4)
tab_model(glm.4)

### Richness - all types

glm.0 <- glm.nb(Observed ~ 1, data=df.div)
glm.1 <- glm.nb(Observed ~ type, data=df.div)
glm.2 <- glm.nb(Observed ~ type + season + site, data=df.div)
glm.3 <- glm.nb(Observed ~ type + season * site, data=df.div)
glm.4 <- glm.nb(Observed ~ type * season + site, data=df.div)
glm.5 <- glm.nb(Observed ~ type * season * site, data=df.div)


AICc(glm.0, glm.1, glm.2, glm.3, glm.4, glm.5)
#glm.1 best (862.2)
summary(glm.1)

#             Estimate Std. Error z value Pr(>|z|)    
#(Intercept)    4.7169     0.1099  42.902  < 2e-16 ***
#typeseawater   0.6658     0.2315   2.876  0.00403 ** 
#typesediment   1.4422     0.2311   6.241 4.36e-10 ***

tab_model(glm.1)
```

```{r shannon glm coral}

# Notes on alpha diversity indices:
    # Shannon - measures richness, so how many different bacteria in each sample
    # Chao1 - richness can be sensitive to read depth; Chao1 has some corrections to overcome this
    # Simpson - measures probability of resampling the same species on two consecutive draws with replacement

df.div.coral$season <- as.factor(df.div.coral$season)
df.div.coral$season <- relevel(df.div.coral$season, ref="Summer")
df.div.coral$site <- as.factor(df.div.coral$site)
df.div.coral$site <- relevel(df.div.coral$site, ref="NR")
df.div.coral$density <- as.numeric(df.div.coral$density)
df.div.coral$density_sqrt <- sqrt(df.div.coral$density)
df.div.coral$density_sqrt <- as.numeric(df.div.coral$density_sqrt)
str(df.div.coral)

qqPlot(df.div.coral$Shannon)
shapiro.test(df.div.coral$Shannon) #normally distributed
ggdensity(df.div.coral$Shannon)

### GLM - Shannon ###
#using Gaussian distribution bc normal distribution

glm.0 <- glm(Shannon ~ 1, data=df.div.coral, family=gaussian())
glm.1a <- glm(Shannon ~ site, data=df.div.coral, family=gaussian())
glm.1b <- glm(Shannon ~ season, data=df.div.coral, family=gaussian())
glm.1c <- glm(Shannon ~ density_sqrt, data=df.div.coral, family=gaussian())

glm.2a <- glm(Shannon ~ season+site, data=df.div.coral, family=gaussian())
glm.2b <- glm(Shannon ~ season+density_sqrt, data=df.div.coral, family=gaussian())
glm.2c <- glm(Shannon ~ site+density_sqrt, data=df.div.coral, family=gaussian())

glm.3a <- glm(Shannon ~ season+site+density_sqrt, data=df.div.coral, family=gaussian())
glm.3b <- glm(Shannon ~ season+site*density_sqrt, data=df.div.coral, family=gaussian())
glm.3c <- glm(Shannon ~ season*site+density_sqrt, data=df.div.coral, family=gaussian())
glm.3d <- glm(Shannon ~ site+season*density_sqrt, data=df.div.coral, family=gaussian())
glm.3e <- glm(Shannon ~ season*site*density_sqrt, data=df.div.coral, family=gaussian())

AICc(glm.0,glm.1a,glm.1b,glm.1c, glm.2a,glm.2b,glm.2c, glm.3a, glm.3b, glm.3c,glm.3d,glm.3e)

#glm.3b best AIC (95.9)
summary(glm.3b)
#                      Estimate Std. Error t value Pr(>|t|)    
#(Intercept)          2.6192306  0.3011504   8.697 1.42e-10 ***
#seasonFall           0.3074556  0.2174935   1.414 0.165611    
#siteRI               1.1748559  0.3747495   3.135 0.003308 ** 
#siteCL               1.6891499  0.5996075   2.817 0.007646 ** 
#density_sqrt         0.0004923  0.0001217   4.047 0.000246 ***
#siteRI:density_sqrt -0.0006796  0.0001524  -4.460 7.07e-05 ***
#siteCL:density_sqrt -0.0010198  0.0002223  -4.588 4.76e-05 ***

1-pchisq(29.785-15.253, 44-38)
#0.0242263 cool

#setwd("~/oculina/diversity/stats")
tab_model(glm.3b, digits=4)
#tab_model(glm.2, file="./stats/sha.coral.glm.gauss.html")
#library(webshot)
#webshot("./stats/sha.coral.glm.gauss.html", "./stats/sha.coral.glm.gauss.png")

# remove outlier from density
library(outliers)
outlier <- outlier(df.div.coral$density_sqrt)
df.div.coral.out <- df.div.coral %>% filter(!density_sqrt %in% outlier)


glm.0 <- glm(Shannon ~ 1, data=df.div.coral.out, family=gaussian())
glm.1a <- glm(Shannon ~ site, data=df.div.coral.out, family=gaussian())
glm.1b <- glm(Shannon ~ season, data=df.div.coral.out, family=gaussian())
glm.1c <- glm(Shannon ~ density_sqrt, data=df.div.coral.out, family=gaussian())

glm.2a <- glm(Shannon ~ season+site, data=df.div.coral.out, family=gaussian())
glm.2b <- glm(Shannon ~ season+density_sqrt, data=df.div.coral.out, family=gaussian())
glm.2c <- glm(Shannon ~ site+density_sqrt, data=df.div.coral.out, family=gaussian())

glm.3a <- glm(Shannon ~ season+site+density_sqrt, data=df.div.coral.out, family=gaussian())
glm.3b <- glm(Shannon ~ season+site*density_sqrt, data=df.div.coral.out, family=gaussian())
glm.3c <- glm(Shannon ~ season*site+density_sqrt, data=df.div.coral.out, family=gaussian())
glm.3d <- glm(Shannon ~ site+season*density_sqrt, data=df.div.coral.out, family=gaussian())
glm.3e <- glm(Shannon ~ season*site*density_sqrt, data=df.div.coral.out, family=gaussian())

AICc(glm.0,glm.1a,glm.1b,glm.1c, glm.2a,glm.2b,glm.2c, glm.3a, glm.3b, glm.3c,glm.3d,glm.3e)
#glm 3b still best
summary(glm.3b)


#trying with symbstate instead of density:

glm.0 <- glm(Shannon ~ 1, data=df.div.coral, family=gaussian())
glm.1a <- glm(Shannon ~ site, data=df.div.coral, family=gaussian())
glm.1b <- glm(Shannon ~ season, data=df.div.coral, family=gaussian())
glm.1c <- glm(Shannon ~ symbstate, data=df.div.coral, family=gaussian())

glm.2a <- glm(Shannon ~ season+site, data=df.div.coral, family=gaussian())
glm.2b <- glm(Shannon ~ season+symbstate, data=df.div.coral, family=gaussian())
glm.2c <- glm(Shannon ~ site+symbstate, data=df.div.coral, family=gaussian())

glm.3a <- glm(Shannon ~ season+site+symbstate, data=df.div.coral, family=gaussian())
glm.3b <- glm(Shannon ~ season+site*symbstate, data=df.div.coral, family=gaussian())
glm.3c <- glm(Shannon ~ season*site+symbstate, data=df.div.coral, family=gaussian())
glm.3d <- glm(Shannon ~ site+season*symbstate, data=df.div.coral, family=gaussian())
glm.3e <- glm(Shannon ~ season*site*symbstate, data=df.div.coral, family=gaussian())

AICc(glm.0,glm.1a,glm.1b,glm.1c, glm.2a,glm.2b,glm.2c, glm.3a, glm.3b, glm.3c,glm.3d,glm.3e)
#glm.3b best AIC (109.6)
summary(glm.3b)
#                           Estimate Std. Error t value Pr(>|t|)    
#(Intercept)                3.119842   0.268222  11.632 4.35e-14 ***
#seasonFall                 0.324202   0.244887   1.324  0.19345    
#siteRI                     0.611807   0.349167   1.752  0.08781 .  
#siteCL                    -0.003359   0.416604  -0.008  0.99361    
#symbstateSymbiotic         0.924520   0.356398   2.594  0.01340 *  
#siteRI:symbstateSymbiotic -1.567137   0.484216  -3.236  0.00251 ** 
#siteCL:symbstateSymbiotic -1.268982   0.574674  -2.208  0.03334 *  

tab_model(glm.3b, digits=2)
tab_model(glm.2, file="./stats/sha.coral.glm.gauss.sqrt.html")
#webshot("./stats/sha.coral.glm.gauss.sqrt.html", "./stats/sha.coral.glm.gauss.sqrt.png")
```


```{r GLM Evenness}
## GLM - Evenness ###
qqPlot(df.div.coral$even)
shapiro.test(df.div.coral$even) #p=0.045
ggdensity(df.div.coral$even) #left skewed?

glm.0 <- glm(even ~ 1, data=df.div.coral, family=gaussian())
glm.1a <- glm(even ~ site, data=df.div.coral, family=gaussian())
glm.1b <- glm(even ~ season, data=df.div.coral, family=gaussian())
glm.1c <- glm(even ~ density_sqrt, data=df.div.coral, family=gaussian())

glm.2a <- glm(even ~ season+site, data=df.div.coral, family=gaussian())
glm.2b <- glm(even ~ season+density_sqrt, data=df.div.coral, family=gaussian())
glm.2c <- glm(even ~ site+density_sqrt, data=df.div.coral, family=gaussian())

glm.3a <- glm(even ~ season+site+density_sqrt, data=df.div.coral, family=gaussian())
glm.3b <- glm(even ~ season+site*density_sqrt, data=df.div.coral, family=gaussian())
glm.3c <- glm(even ~ season*site+density_sqrt, data=df.div.coral, family=gaussian())
glm.3d <- glm(even ~ site+season*density_sqrt, data=df.div.coral, family=gaussian())
glm.3e <- glm(even ~ season*site*density_sqrt, data=df.div.coral, family=gaussian())

AICc(glm.0,glm.1a,glm.1b,glm.1c, glm.2a,glm.2b,glm.2c, glm.3a, glm.3b, glm.3c,glm.3d,glm.3e)

# glm.1b best (-88.0)
summary(glm.1b) 

tab_model(glm.1b, digits=2)

### Symbstate - evenness

glm.0 <- glm(even ~ 1, data=df.div.coral, family=gaussian())
glm.1a <- glm(even ~ site, data=df.div.coral, family=gaussian())
glm.1b <- glm(even ~ season, data=df.div.coral, family=gaussian())
glm.1c <- glm(even ~ symbstate, data=df.div.coral, family=gaussian())

glm.2a <- glm(even ~ season+site, data=df.div.coral, family=gaussian())
glm.2b <- glm(even ~ season+symbstate, data=df.div.coral, family=gaussian())
glm.2c <- glm(even ~ site+symbstate, data=df.div.coral, family=gaussian())

glm.3a <- glm(even ~ season+site+symbstate, data=df.div.coral, family=gaussian())
glm.3b <- glm(even ~ season+site*symbstate, data=df.div.coral, family=gaussian())
glm.3c <- glm(even ~ season*site+symbstate, data=df.div.coral, family=gaussian())
glm.3d <- glm(even ~ site+season*symbstate, data=df.div.coral, family=gaussian())
glm.3e <- glm(even ~ season*site*symbstate, data=df.div.coral, family=gaussian())

AICc(glm.0,glm.1a,glm.1b,glm.1c, glm.2a,glm.2b,glm.2c, glm.3a, glm.3b, glm.3c,glm.3d,glm.3e)
 #glm.1b best (-88.0)

summary(glm.1b)
tab_model(glm.1b)
```



```{r GLM richness}
## GLM - richness ###
qqPlot(df.div.coral$Observed) # discrete, not continuous
shapiro.test(df.div.coral$Observed) #really sig/not normal
ggdensity(df.div.coral$Observed) #right skewed

#using poisson because counts; however, poisson is sensitive to assumption that mean is equal to variance; alternative is quasi-poisson or negative binomial
glm.0 <- glm(Observed ~ 1, data=df.div.coral, family=poisson())
glm.1 <- glm(Observed ~ season+site+density, data=df.div.coral, family=poisson())
#summary(glm.1)
glm.2 <- glm(Observed ~ season+site*density, data=df.div.coral, family=poisson())
#summary(glm.2)
glm.3 <- glm(Observed ~ season*site+density, data=df.div.coral, family=poisson())
#summary(glm.3)
glm.4 <- glm(Observed ~ site+season*density, data=df.div.coral, family=poisson())
#summary(glm.4)
glm.5 <- glm(Observed ~ season*site*density, data=df.div.coral, family=poisson())
#summary(glm.5)

AICc(glm.0,glm.1, glm.2, glm.3, glm.4, glm.5)
# glm.5 is best (1502.1)
summary(glm.5) 

#for glm.5:
#Coefficients: (2 not defined because of singularities)
#^^^why does this happen?! i think bc of CL...can I change CL to not be intercept?


# trying negative binomial
library(MASS)


glm.0 <- glm.nb(Observed ~ 1, data=df.div.coral)
glm.1a <- glm.nb(Observed ~ site, data=df.div.coral)
glm.1b <- glm.nb(Observed ~ season, data=df.div.coral)
glm.1c <- glm.nb(Observed ~ density_sqrt, data=df.div.coral)

glm.2a <- glm.nb(Observed ~ season+site, data=df.div.coral)
glm.2b <- glm.nb(Observed ~ season+density_sqrt, data=df.div.coral)
glm.2c <- glm.nb(Observed ~ site+density_sqrt, data=df.div.coral)

glm.3a <- glm.nb(Observed ~ season+site+density_sqrt, data=df.div.coral)
glm.3b <- glm.nb(Observed ~ season+site*density_sqrt, data=df.div.coral)
glm.3c <- glm.nb(Observed ~ season*site+density_sqrt, data=df.div.coral)
glm.3d <- glm.nb(Observed ~ site+season*density_sqrt, data=df.div.coral)
glm.3e <- glm.nb(Observed ~ season*site*density_sqrt, data=df.div.coral)

AICc(glm.0,glm.1a,glm.1b,glm.1c, glm.2a,glm.2b,glm.2c, glm.3a, glm.3b, glm.3c,glm.3d,glm.3e)

# glm.3b best

summary(glm.3b)
#                      Estimate Std. Error z value Pr(>|z|)    
#(Intercept)          3.909e+00  2.423e-01  16.130  < 2e-16 ***
#seasonFall          -1.973e-01  1.734e-01  -1.138 0.255077    
#siteRI               1.297e+00  3.012e-01   4.307 1.66e-05 ***
#siteCL               1.805e+00  4.779e-01   3.778 0.000158 ***
#density_sqrt         4.820e-04  9.724e-05   4.957 7.17e-07 ***
#siteRI:density_sqrt -7.518e-04  1.224e-04  -6.144 8.06e-10 ***
#siteCL:density_sqrt -1.035e-03  1.781e-04  -5.810 6.26e-09 ***

tab_model(glm.3b, digits=4)


### symbiotic state
glm.0 <- glm.nb(Observed ~ 1, data=df.div.coral)
glm.1a <- glm.nb(Observed ~ site, data=df.div.coral)
glm.1b <- glm.nb(Observed ~ season, data=df.div.coral)
glm.1c <- glm.nb(Observed ~ symbstate, data=df.div.coral)

glm.2a <- glm.nb(Observed ~ season+site, data=df.div.coral)
glm.2b <- glm.nb(Observed ~ season+symbstate, data=df.div.coral)
glm.2c <- glm.nb(Observed ~ site+symbstate, data=df.div.coral)

glm.3a <- glm.nb(Observed ~ season+site+symbstate, data=df.div.coral)
glm.3b <- glm.nb(Observed ~ season+site*symbstate, data=df.div.coral)
glm.3c <- glm.nb(Observed ~ season*site+symbstate, data=df.div.coral)
glm.3d <- glm.nb(Observed ~ site+season*symbstate, data=df.div.coral)
glm.3e <- glm.nb(Observed ~ season*site*symbstate, data=df.div.coral)

AICc(glm.0,glm.1a,glm.1b,glm.1c, glm.2a,glm.2b,glm.2c, glm.3a, glm.3b, glm.3c,glm.3d,glm.3e) #null model best, 3b is close second

summary(glm.3b)
tab_model(glm.3b)
```

```{r GLM Simpson}
## GLM - Simpson ###
qqPlot(df.div.coral$Simpson)
shapiro.test(df.div.coral$Simpson) #not normally distributed
ggdensity(df.div.coral$Simpson) #very left skewed

#maybe should use gamma? https://stats.stackexchange.com/questions/190763/how-to-decide-which-glm-family-to-use


glm.0 <- glm(Simpson ~ 1, data=df.div.coral, family=Gamma())
glm.1 <- glm(Simpson ~ season+site+density, data=df.div.coral, family=Gamma())
#summary(glm.1)
glm.2 <- glm(Simpson ~ season+site*density, data=df.div.coral, family=Gamma())
#summary(glm.2)
glm.3 <- glm(Simpson ~ season*site+density, data=df.div.coral, family=Gamma())
#summary(glm.3)
glm.4 <- glm(Simpson ~ site+season*density, data=df.div.coral, family=Gamma())
#summary(glm.4)
glm.5 <- glm(Simpson ~ season*site*density, data=df.div.coral, family=Gamma())
#summary(glm.5)

AICc(glm.0,glm.1, glm.2, glm.3, glm.4, glm.5)
# glm.4 is best, glm.0 close second
summary(glm.4) #for gamma
#                       Estimate Std. Error t value Pr(>|t|)    
#(Intercept)           1.138e+00  4.941e-02  23.031   <2e-16 ***
#siteNR               -9.423e-02  4.100e-02  -2.298   0.0270 *  
#siteRI               -7.796e-02  4.187e-02  -1.862   0.0702 .  
#seasonSummer          3.905e-02  4.249e-02   0.919   0.3637    
#density              -1.525e-09  5.020e-09  -0.304   0.7629    
#seasonSummer:density  2.289e-09  5.575e-09   0.411   0.6836   

#library(lmtest)
#check for heteroskedasticity
bptest(glm.4)
```

```{r GLM InvSimpson}
## GLM - InvSimpson ###
qqPlot(df.div.coral$InvSimpson)
shapiro.test(df.div.coral$InvSimpson) #not normally distributed
ggdensity(df.div.coral$InvSimpson) #very right skewed - I think gamma best

#models:
glm.0 <- glm(InvSimpson ~ 1, data=df.div.coral, family=Gamma())
glm.1 <- glm(InvSimpson ~ season+site+density, data=df.div.coral, family=Gamma())
#summary(glm.1)
glm.2 <- glm(InvSimpson ~ season+site*density, data=df.div.coral, family=Gamma())
#summary(glm.2)
glm.3 <- glm(InvSimpson ~ season*site+density, data=df.div.coral, family=Gamma())
#summary(glm.3)
glm.4 <- glm(InvSimpson ~ site+season*density, data=df.div.coral, family=Gamma())
#summary(glm.4)
glm.5 <- glm(InvSimpson ~ season*site*density, data=df.div.coral, family=Gamma())
#summary(glm.5)

AICc(glm.1, glm.2, glm.3, glm.4, glm.5)
# glm.2 best
summary(glm.2)
#                 Estimate Std. Error t value Pr(>|t|)    
#(Intercept)     1.110e-02  1.671e-02   0.664 0.510461    
#seasonSummer    1.016e-02  7.965e-03   1.276 0.209730    
#siteNR          5.823e-02  2.197e-02   2.650 0.011670 *  
#siteRI          1.202e-02  1.708e-02   0.704 0.485868    
#density         1.101e-08  3.793e-09   2.904 0.006115 ** 
#siteNR:density -1.536e-08  4.052e-09  -3.790 0.000523 ***
#siteRI:density -8.700e-09  3.985e-09  -2.183 0.035290 *   

#Coefficients: (2 not defined because of singularities)
#^^^Why?!
```

```{r GLM excluding CL - sanity check?}
df.div.coral.NR.RI$season <- as.factor(df.div.coral.NR.RI$season)
df.div.coral.NR.RI$site <- as.factor(df.div.coral.NR.RI$site)
df.div.coral.NR.RI$density <- as.numeric(df.div.coral.NR.RI$density)
str(df.div.coral.NR.RI)

glm.0 <- glm(Shannon ~ 1, data=df.div.coral.NR.RI, family=gaussian())
glm.1 <- glm(Shannon ~ season+site+density, data=df.div.coral.NR.RI, family=gaussian())
#summary(glm.1)
glm.2 <- glm(Shannon ~ season+site*density, data=df.div.coral.NR.RI, family=gaussian())
#summary(glm.2)
glm.3 <- glm(Shannon ~ season*site+density, data=df.div.coral.NR.RI, family=gaussian())
#summary(glm.3)
glm.4 <- glm(Shannon ~ site+season*density, data=df.div.coral.NR.RI, family=gaussian())
#summary(glm.4)
glm.5 <- glm(Shannon ~ season*site*density, data=df.div.coral.NR.RI, family=gaussian())

```

### Faith's D


## Phylogenetic diversity (Faith's D)

Tutorial from dada2 author [here](https://f1000research.com/articles/5-1492/v2)

```{r packages phylo d, eval=FALSE}
#install.packages('devtools')
#library(devtools)
#devtools::install_github('twbattaglia/btools')
#library(btools)
```


(I'm not running the following chunk every time, because only need to generate the file once)

```{r fasta file generation, eval=FALSE}
raw.otu  <- as.matrix(ps.less@otu_table)
raw.taxa <- data.frame(ps.less@tax_table)
rownames(raw.taxa)==colnames(raw.otu)
 
colnames(raw.otu) <- raw.taxa$V8
ids <- rownames(raw.taxa)

path="~/oculina/diversity/analyze_asv_table/oculina16s_rev.less.fasta"
uniquesToFasta(raw.otu, path, ids = ids, mode = "w", width = 20000)
```


Actual analysis part: 

(I'm not running the following chunk every time, because only need to generate the files once)

```{r phylo d, eval=FALSE}
seqs <- getSequences("oculina16s_rev.less.fasta")
names(seqs) <- seqs # This propagates to the tip labels of the tree
saveRDS(seqs,file="phylo.seqs.less.rds")
#also doing the same thing with a .fasta file post-trimming to see if it makes a difference:
#seqs <- getSequences("mr16s_rev.cleanest.trimmed copy.fasta")
#names(seqs) <- seqs # This propagates to the tip labels of the tree
#saveRDS(seqs,file="phylo.seqs.rev.cleanest.trimmed.rds")
```


Doing this next part in the cluster because it takes forever

```{bash terminal phylo things, eval=FALSE}
##script phylo.R looks like this now:
# library(dada2)
# library(phangorn)
# library(DECIPHER)
#
# seqs <- readRDS("./phylo.seqs.rev.cleanest.rds")
# alignment <- AlignSeqs(DNAStringSet(seqs), anchor=NA)
# phang.align <- phyDat(as(alignment, "matrix"), type="DNA")
# dm <- dist.ml(phang.align)
# treeNJ <- NJ(dm) # Note, tip order != sequence order
# fit = pml(treeNJ, data=phang.align)
#
# ## negative edges length changed to 0!
# 
# fitGTR <- update(fit, k=4, inv=0.2)
# fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
#                     rearrangement = "stochastic", control = pml.control(trace = 0))
# detach("package:phangorn", unload=TRUE)
# saveRDS(fitGTR, file="./phylo.fitgtr.rev.cleanest.rds")

### In terminal:
#nano phylo.sh #to create and edit text file
#phylo.sh looks like this: 
  # #!/bin/bash
  # #SBATCH -p general
  # #SBATCH -t 1-
  # #SBATCH -N 1
  # #SBATCH -n 1
  # #SBATCH --mem=8gb
  # #SBATCH --output <phylo.fitgtr.less.rds>
  
  # module add r
  # module add rstudio
  
  # Rscript phylo.R

##saved output as: phylo.fitgtr.less.rds
```

Back in R
```{r}
#BiocManager::install("DESeq2")
#BiocManager::install("genefilter")
library(DESeq2)
library(genefilter)
#devtools::install_github('twbattaglia/btools')
library(btools)
setwd("~/oculina/diversity/analyze_asv_table")
fitGTR <- readRDS("phylo.fitgtr.less.rds")

#new phyloseq object:
taxa.less <- data.frame(ps.less@tax_table)
seqtab.less <- data.frame(ps.less@otu_table)
taxa.less$sqs <- row.names(taxa.less) 
taxa.less$sqs == colnames(seqtab.less)
row.names(taxa.less) <- taxa.less$V8
colnames(seqtab.less) <- taxa.less$V8
row.names(taxa.less) == colnames(seqtab.less)
taxa.less <- as.matrix(taxa.less)
ps.less.tree <- phyloseq(otu_table(seqtab.less, taxa_are_rows = FALSE),
                         sample_data(samdf),
                         tax_table(taxa.less),
                         phy_tree(fitGTR$tree))

pd.div <- estimate_pd(ps.less.tree)
row.names(df.div) <- df.div$id
df.div.pd <- merge(df.div,pd.div,by=0)

#saveRDS(ps.less.tree,file="ps.less.tree.Rdata")
## saving diversity data frame ##
#save & read back in as needed
#write.csv(df.div.pd,file="oculina16s_diversity_less.csv") #saving
#df.div <- read.csv("oculina16s_diversity_less.csv",row.names=1,header=TRUE) #reading back in

```


```{r faith's d GLMs}
#setwd("~oculina/data")
ps.less.tree <- readRDS("ps.less.tree.Rdata")
pd.div <- estimate_pd(ps.less.tree)
row.names(df.div) <- df.div$id
df.div.pd <- merge(df.div,pd.div,by=0)
df.div.coral.pd <- subset(df.div.pd, type=="coral")
df.div.coral.pd$density <- as.numeric(df.div.coral.pd$density)
df.div.coral.pd$density_sqrt <- sqrt(df.div.coral.pd$density)
df.div.coral.pd$season <- as.factor(df.div.coral$season)
df.div.coral.pd$season <- relevel(df.div.coral.pd$season, ref="Summer")
df.div.coral.pd$site <- as.factor(df.div.coral.pd$site)
df.div.coral.pd$site <- relevel(df.div.coral.pd$site, ref="NR")
str(df.div.coral.pd)


## GLM - Faith's D ###
# All types
df.div.pd$season <- as.factor(df.div.pd$season)
df.div.pd$site <- as.factor(df.div.pd$site)
df.div.pd$type <- as.factor(df.div.pd$type)

glm.0 <- glm(PD ~ 1, data=df.div.pd, family=gaussian())
glm.1 <- glm(PD ~ type, data=df.div.pd, family=gaussian)
glm.2 <- glm(PD ~ type + season + site, data=df.div.pd, family=gaussian)
glm.3 <- glm(PD ~ type + season * site, data=df.div.pd, family=gaussian)
glm.4 <- glm(PD ~ type * season + site, data=df.div.pd, family=gaussian)
glm.5 <- glm(PD ~ type * season * site, data=df.div.pd, family=gaussian)

AICc(glm.0, glm.1, glm.2, glm.3, glm.4, glm.5)
#glm.5 best by a smidgen, null model second best (AIC 706.1)
summary(glm.5)

#qqPlot(df.div.coral.pd$PD)
#shapiro.test(df.div.coral.pd$PD) #p=0.0011
#ggdensity(df.div.coral.pd$PD) #right-skewed

glm.0 <- glm(PD ~ 1, data=df.div.coral.pd, family=gaussian())
glm.1a <- glm(PD ~ site, data=df.div.coral.pd, family=gaussian())
glm.1b <- glm(PD ~ season, data=df.div.coral.pd, family=gaussian())
glm.1c <- glm(PD ~ density_sqrt, data=df.div.coral.pd, family=gaussian())

glm.2a <- glm(PD ~ season+site, data=df.div.coral.pd, family=gaussian())
glm.2b <- glm(PD ~ season+density_sqrt, data=df.div.coral.pd, family=gaussian())
glm.2c <- glm(PD ~ site+density_sqrt, data=df.div.coral.pd, family=gaussian())

glm.3a <- glm(PD ~ season+site+density_sqrt, data=df.div.coral.pd, family=gaussian())
glm.3b <- glm(PD ~ season+site*density_sqrt, data=df.div.coral.pd, family=gaussian())
glm.3c <- glm(PD ~ season*site+density_sqrt, data=df.div.coral.pd, family=gaussian())
glm.3d <- glm(PD ~ site+season*density_sqrt, data=df.div.coral.pd, family=gaussian())
glm.3e <- glm(PD ~ season*site*density_sqrt, data=df.div.coral.pd, family=gaussian())

AICc(glm.0,glm.1a,glm.1b,glm.1c, glm.2a,glm.2b,glm.2c, glm.3a, glm.3b, glm.3c,glm.3d,glm.3e)

# glm.3b best
summary(glm.3b) 

tab_model(glm.3b, digits=2)

### Symbstate - Faith's D
glm.0 <- glm(PD ~ 1, data=df.div.coral.pd, family=gaussian())
glm.1a <- glm(PD ~ site, data=df.div.coral.pd, family=gaussian())
glm.1b <- glm(PD ~ season, data=df.div.coral.pd, family=gaussian())
glm.1c <- glm(PD ~ symbstate, data=df.div.coral.pd, family=gaussian())

glm.2a <- glm(PD ~ season+site, data=df.div.coral.pd, family=gaussian())
glm.2b <- glm(PD ~ season+symbstate, data=df.div.coral.pd, family=gaussian())
glm.2c <- glm(PD ~ site+symbstate, data=df.div.coral.pd, family=gaussian())

glm.3a <- glm(PD ~ season+site+symbstate, data=df.div.coral.pd, family=gaussian())
glm.3b <- glm(PD ~ season+site*symbstate, data=df.div.coral.pd, family=gaussian())
glm.3c <- glm(PD ~ season*site+symbstate, data=df.div.coral.pd, family=gaussian())
glm.3d <- glm(PD ~ site+season*symbstate, data=df.div.coral.pd, family=gaussian())
glm.3e <- glm(PD ~ season*site*symbstate, data=df.div.coral.pd, family=gaussian())

AICc(glm.0,glm.1a,glm.1b,glm.1c, glm.2a,glm.2b,glm.2c, glm.3a, glm.3b, glm.3c,glm.3d,glm.3e)
 #null model best, glm.1a second best but no significance for sites
summary(glm.1a)

tab_model(glm.3b)
```

#summary(glm.5)

AICc(glm.0,glm.1, glm.2, glm.3, glm.4, glm.5)
#glm.2 best model (still! woo!)

summary(glm.2)
#                Estimate Std. Error t value Pr(>|t|)    
#(Intercept)     3.211662   0.271487  11.830 7.97e-13 ***
#seasonSummer   -0.267047   0.227255  -1.175 0.249198    
#siteRI          0.650927   0.302829   2.149 0.039783 *  
#density         0.018828   0.004934   3.816 0.000632 ***
#siteRI:density -0.022815   0.005573  -4.094 0.000295 ***
```

